core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/chunkedencoding/AwsSignedChunkedEncodingInputStream.java

```diff
@@ -0,0 +1,224 @@
+ * the wrapped stream.
+ */
+@SdkProtectedApi
+public final class AwsSignedChunkedEncodingInputStream extends AwsChunkedEncodingInputStream {
+
+    private static final String CRLF = "\r\n";
+    private static final String CHUNK_SIGNATURE_HEADER = ";chunk-signature=";
+    private static final String CHECKSUM_SIGNATURE_HEADER = "x-amz-trailer-signature:";
+    private final AwsChunkSigner chunkSigner;
+    private final String headerSignature;
+    private String previousChunkSignature;
+
+    /**
+     * Creates a chunked encoding input stream initialized with the originating stream, an http request seed signature
+     * and a signer that can sign a chunk of bytes according to a chosen algorithm. The configuration allows
+     * specification of the size of each chunk, as well as the buffer size.
+     *
+     * @param in              The original InputStream.
+     * @param headerSignature The signature of the signed headers of the request. This will be used for
+     *                        calculating the signature of the first chunk. Observe that the format of
+     *                        this parameter should be a hex-encoded string.
+     * @param chunkSigner     The signer for each chunk of data, implementing the {@link AwsChunkSigner} interface.
+     * @param config          The configuration allows the user to customize chunk size and buffer size.
+     *                        See {@link AwsChunkedEncodingConfig} for default values.
+     */
+    private AwsSignedChunkedEncodingInputStream(InputStream in, SdkChecksum sdkChecksum,
+                                                String checksumHeaderForTrailer,
+                                                String headerSignature,
+                                                AwsChunkSigner chunkSigner,
+                                                AwsChunkedEncodingConfig config) {
+        super(in, sdkChecksum, checksumHeaderForTrailer, config);
+        this.chunkSigner = chunkSigner;
+        this.previousChunkSignature = headerSignature;
+        this.headerSignature = headerSignature;
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    /**
+     * Calculates the expected total length of signed payload chunked stream.
+     *
+     * @param originalLength  The length of the data
+     * @param signatureLength The length of a calculated signature, dependent on which {@link AwsChunkSigner} is used
+     * @param config          The chunked encoding config determines the size of the chunks. Use the same values as when
+     *                        initializing the stream.
+     */
+    public static long calculateStreamContentLength(long originalLength,
+                                                    int signatureLength,
+                                                    AwsChunkedEncodingConfig config,
+                                                    boolean isTrailingChecksumCalculated) {
```

Comment:
This feels brittle for a protected API. Let's reassess once checksums are out, but protected APIs need to focus on forward-compatibility, or they're better off just being copied and treated as internal APIs. This feels like it only suits today's needs.